File: Common_Assignment/js/compiler/lexer.js
/**
 * ScopeLab - Lexer (Lexical Analyzer)
 */

import { Token, TokenType, KEYWORDS, TYPE_KEYWORDS } from './tokens.js';

export class Lexer {
    constructor(source) {
        this.source = source || '';
        this.cursor = 0;
        this.line = 1;
        this.column = 1;
        this.tokens = [];
        this.errors = [];
    }

    tokenize() {
        this.tokens = [];
        this.errors = [];
        this.cursor = 0;
        this.line = 1;
        this.column = 1;

        while (!this.isAtEnd()) {
            this.skipWhitespaceAndComments();
            if (this.isAtEnd()) break;

            const char = this.peek();
            const startLine = this.line;
            const startCol = this.column;

            // Numbers (int or float)
            if (this.isDigit(char)) {
                this.readNumber();
            }
            // Strings
            else if (char === '"' || char === "'") {
                this.readString(char);
            }
            // Identifiers / Keywords / Types
            else if (this.isAlpha(char) || char === '_') {
                this.readIdentifierOrKeyword();
            }
            // Operators & Delimiters
            else {
                this.readSymbol();
            }
        }

        this.tokens.push(new Token(TokenType.EOF, '', this.line, this.column, 'special'));
        return {
            tokens: this.tokens,
            errors: this.errors
        };
    }

    isAtEnd() {
        return this.cursor >= this.source.length;
    }

    peek(offset = 0) {
        if (this.cursor + offset >= this.source.length) return '\0';
        return this.source[this.cursor + offset];
    }

    advance() {
        if (this.isAtEnd()) return '\0';
        const char = this.source[this.cursor++];
        if (char === '\n') {
            this.line++;
            this.column = 1;
        } else {
            this.column++;
        }
        return char;
    }

    match(expected) {
        if (this.isAtEnd()) return false;
        if (this.source[this.cursor] !== expected) return false;
        this.advance();
        return true;
    }

    skipWhitespaceAndComments() {
        while (!this.isAtEnd()) {
            const char = this.peek();

            if (char === ' ' || char === '\t' || char === '\r' || char === '\n') {
                this.advance();
            }
            // Single-line comment
            else if (char === '/' && this.peek(1) === '/') {
                this.advance(); // consume '/'
                this.advance(); // consume '/'
                while (!this.isAtEnd() && this.peek() !== '\n') {
                    this.advance();
                }
            }
            // Multi-line comment
            else if (char === '/' && this.peek(1) === '*') {
                this.advance(); // consume '/'
                this.advance(); // consume '*'
                while (!this.isAtEnd()) {
                    if (this.peek() === '*' && this.peek(1) === '/') {
                        this.advance(); // consume '*'
                        this.advance(); // consume '/'
                        break;
                    }
                    this.advance();
                }
            } else {
                break;
            }
        }
    }

    isDigit(char) {
        return char >= '0' && char <= '9';
    }

    isAlpha(char) {
        return (char >= 'a' && char <= 'z') ||
               (char >= 'A' && char <= 'Z');
    }

    isAlphaNumeric(char) {
        return this.isAlpha(char) || this.isDigit(char) || char === '_';
    }

    readNumber() {
        const startLine = this.line;
        const startCol = this.column;
        let numStr = '';
        let isFloat = false;

        while (this.isDigit(this.peek())) {
            numStr += this.advance();
        }

        // Check for fractional part
        if (this.peek() === '.' && this.isDigit(this.peek(1))) {
            isFloat = true;
            numStr += this.advance(); // consume '.'
            while (this.isDigit(this.peek())) {
                numStr += this.advance();
            }
        }

        const type = isFloat ? TokenType.FLOAT_LITERAL : TokenType.INT_LITERAL;
        this.tokens.push(new Token(type, numStr, startLine, startCol, isFloat ? 'float' : 'integer'));
    }

    readString(quoteChar) {
        const startLine = this.line;
        const startCol = this.column;
        this.advance(); // consume opening quote
        let strVal = '';

        while (!this.isAtEnd() && this.peek() !== quoteChar && this.peek() !== '\n') {
            if (this.peek() === '\\') {
                this.advance();
                const escaped = this.advance();
                if (escaped === 'n') strVal += '\n';
                else if (escaped === 't') strVal += '\t';
                else if (escaped === '"') strVal += '"';
                else if (escaped === "'") strVal += "'";
                else strVal += escaped;
            } else {
                strVal += this.advance();
            }
        }

        if (this.peek() === quoteChar) {
            this.advance(); // consume closing quote
            this.tokens.push(new Token(TokenType.STRING_LITERAL, strVal, startLine, startCol, 'string'));
        } else {
            this.errors.push({
                type: 'LEXICAL_ERROR',
                message: `Unterminated string literal starting at line ${startLine}`,
                line: startLine,
                column: startCol
            });
            this.tokens.push(new Token(TokenType.STRING_LITERAL, strVal, startLine, startCol, 'string'));
        }
    }

    readIdentifierOrKeyword() {
        const startLine = this.line;
        const startCol = this.column;
        let word = '';

        while (this.isAlphaNumeric(this.peek())) {
            word += this.advance();
        }

        if (TYPE_KEYWORDS.has(word)) {
            this.tokens.push(new Token(TokenType.TYPE, word, startLine, startCol, 'type'));
        } else if (KEYWORDS.has(word)) {
            this.tokens.push(new Token(TokenType.KEYWORD, word, startLine, startCol, 'keyword'));
        } else {
            this.tokens.push(new Token(TokenType.IDENTIFIER, word, startLine, startCol, 'identifier'));
        }
    }

    readSymbol() {
        const startLine = this.line;
        const startCol = this.column;
        const char = this.advance();

        // 2-character operators
        if (char === '=' && this.peek() === '=') {
            this.advance();
            this.tokens.push(new Token(TokenType.OPERATOR, '==', startLine, startCol, 'operator'));
            return;
        }
        if (char === '!' && this.peek() === '=') {
            this.advance();
            this.tokens.push(new Token(TokenType.OPERATOR, '!=', startLine, startCol, 'operator'));
            return;
        }
        if (char === '<' && this.peek() === '=') {
            this.advance();
            this.tokens.push(new Token(TokenType.OPERATOR, '<=', startLine, startCol, 'operator'));
            return;
        }
        if (char === '>' && this.peek() === '=') {
            this.advance();
            this.tokens.push(new Token(TokenType.OPERATOR, '>=', startLine, startCol, 'operator'));
            return;
        }

        // Single-character operators & delimiters
        switch (char) {
            case '{':
            case '}':
            case '(':
            case ')':
            case ',':
            case ';':
                this.tokens.push(new Token(TokenType.DELIMITER, char, startLine, startCol, 'delimiter'));
                break;
            case '+':
            case '-':
            case '*':
            case '/':
            case '%':
            case '=':
            case '<':
            case '>':
                this.tokens.push(new Token(TokenType.OPERATOR, char, startLine, startCol, 'operator'));
                break;
            default:
                this.errors.push({
                    type: 'LEXICAL_ERROR',
                    message: `Unrecognized character '${char}' at line ${startLine}, column ${startCol}`,
                    line: startLine,
                    column: startCol
                });
                this.tokens.push(new Token(TokenType.UNKNOWN, char, startLine, startCol, 'unknown'));
                break;
        }
    }
}

File: craftlang-compiler/craftlang/lexer/lexer.py
"""Lexer implementation for CraftLang."""

from typing import List, Optional, Any
from .tokens import TokenType, Token, KEYWORDS
from ..errors import LexerError


class Lexer:
    """Scans CraftLang source text into a sequential stream of Tokens."""

    def __init__(self, source: str):
        self.source: str = source
        self.length: int = len(source)
        self.lines: List[str] = source.splitlines()
        self.pos: int = 0
        self.line: int = 1
        self.col: int = 1
        self.tokens: List[Token] = []

    def _get_current_line(self, line_num: int) -> Optional[str]:
        idx = line_num - 1
        if 0 <= idx < len(self.lines):
            return self.lines[idx]
        return None

    def _peek(self, offset: int = 0) -> str:
        idx = self.pos + offset
        if idx < self.length:
            return self.source[idx]
        return '\0'

    def _advance(self) -> str:
        ch = self._peek()
        self.pos += 1
        if ch == '\n':
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def tokenize(self) -> List[Token]:
        """Tokenizes the entire source string and returns a list of Tokens."""
        self.tokens = []
        while self.pos < self.length:
            ch = self._peek()

            # Skip whitespace
            if ch in (' ', '\t', '\r', '\n'):
                self._advance()
                continue

            # Comments
            if ch == '/' and self._peek(1) == '/':
                # Single-line comment
                while self._peek() != '\0' and self._peek() != '\n':
                    self._advance()
                continue

            if ch == '/' and self._peek(1) == '*':
                # Multi-line comment
                start_line, start_col = self.line, self.col
                self._advance()  # '/'
                self._advance()  # '*'
                closed = False
                while self._peek() != '\0':
                    if self._peek() == '*' and self._peek(1) == '/':
                        self._advance()
                        self._advance()
                        closed = True
                        break
                    self._advance()
                if not closed:
                    raise LexerError(
                        "Unterminated multi-line comment",
                        line=start_line,
                        column=start_col,
                        length=2,
                        source_line=self._get_current_line(start_line),
                    )
                continue

            start_line = self.line
            start_col = self.col

            # Numbers (Integer & Float)
            if ch.isdigit():
                self._lex_number(start_line, start_col)
                continue

            # Strings
            if ch == '"':
                self._lex_string(start_line, start_col)
                continue

            # Identifiers and Keywords
            if ch.isalpha() or ch == '_':
                self._lex_identifier(start_line, start_col)
                continue

            # Two-character and Single-character Operators / Punctuation
            self._lex_symbol(start_line, start_col)

        # Append EOF token
        self.tokens.append(Token(
            type=TokenType.EOF,
            value=None,
            raw="",
            line=self.line,
            column=self.col,
            length=0,
        ))
        return self.tokens

    def _lex_number(self, start_line: int, start_col: int) -> None:
        start_pos = self.pos
        is_float = False

        while self._peek().isdigit():
            self._advance()

        # Check for fractional point
        if self._peek() == '.' and self._peek(1).isdigit():
            is_float = True
            self._advance()  # consume '.'
            while self._peek().isdigit():
                self._advance()

        raw = self.source[start_pos:self.pos]
        val = float(raw) if is_float else int(raw)
        tok_type = TokenType.FLOAT_LIT if is_float else TokenType.INT_LIT

        self.tokens.append(Token(
            type=tok_type,
            value=val,
            raw=raw,
            line=start_line,
            column=start_col,
            length=len(raw),
        ))

    def _lex_string(self, start_line: int, start_col: int) -> None:
        start_pos = self.pos
        self._advance()  # Skip opening quote
        chars = []

        while True:
            ch = self._peek()
            if ch == '\0' or ch == '\n':
                raise LexerError(
                    "Unterminated string literal",
                    line=start_line,
                    column=start_col,
                    length=self.pos - start_pos,
                    source_line=self._get_current_line(start_line),
                )
            if ch == '"':
                self._advance()  # Skip closing quote
                break

            if ch == '\\':
                self._advance()
                esc = self._advance()
                if esc == 'n':
                    chars.append('\n')
                elif esc == 't':
                    chars.append('\t')
                elif esc == 'r':
                    chars.append('\r')
                elif esc == '"':
                    chars.append('"')
                elif esc == '\\':
                    chars.append('\\')
                else:
                    chars.append(esc)
            else:
                chars.append(self._advance())

        raw = self.source[start_pos:self.pos]
        val = "".join(chars)
        self.tokens.append(Token(
            type=TokenType.STRING_LIT,
            value=val,
            raw=raw,
            line=start_line,
            column=start_col,
            length=len(raw),
        ))

    def _lex_identifier(self, start_line: int, start_col: int) -> None:
        start_pos = self.pos
        while self._peek().isalnum() or self._peek() == '_':
            self._advance()

        raw = self.source[start_pos:self.pos]
        tok_type = KEYWORDS.get(raw, TokenType.IDENTIFIER)

        val: Any = raw
        if tok_type == TokenType.TRUE:
            tok_type = TokenType.BOOL_LIT
            val = True
        elif tok_type == TokenType.FALSE:
            tok_type = TokenType.BOOL_LIT
            val = False

        self.tokens.append(Token(
            type=tok_type,
            value=val,
            raw=raw,
            line=start_line,
            column=start_col,
            length=len(raw),
        ))

    def _lex_symbol(self, start_line: int, start_col: int) -> None:
        ch = self._advance()

        # Two-char checks
        if ch == '=':
            if self._peek() == '=':
                self._advance()
                self.tokens.append(Token(TokenType.EQ_EQ, "==", "==", start_line, start_col, 2))
                return
            self.tokens.append(Token(TokenType.ASSIGN, "=", "=", start_line, start_col, 1))
            return

        if ch == '!':
            if self._peek() == '=':
                self._advance()
                self.tokens.append(Token(TokenType.BANG_EQ, "!=", "!=", start_line, start_col, 2))
                return
            self.tokens.append(Token(TokenType.BANG, "!", "!", start_line, start_col, 1))
            return

        if ch == '<':
            if self._peek() == '=':
                self._advance()
                self.tokens.append(Token(TokenType.LTE, "<=", "<=", start_line, start_col, 2))
                return
            self.tokens.append(Token(TokenType.LT, "<", "<", start_line, start_col, 1))
            return

        if ch == '>':
            if self._peek() == '=':
                self._advance()
                self.tokens.append(Token(TokenType.GTE, ">=", ">=", start_line, start_col, 2))
                return
            self.tokens.append(Token(TokenType.GT, ">", ">", start_line, start_col, 1))
            return

        if ch == '&' and self._peek() == '&':
            self._advance()
            self.tokens.append(Token(TokenType.AND, "&&", "&&", start_line, start_col, 2))
            return

        if ch == '|' and self._peek() == '|':
            self._advance()
            self.tokens.append(Token(TokenType.OR, "||", "||", start_line, start_col, 2))
            return

        if ch == '-' and self._peek() == '>':
            self._advance()
            self.tokens.append(Token(TokenType.ARROW, "->", "->", start_line, start_col, 2))
            return

        # Single-char symbols
        single_map = {
            '+': (TokenType.PLUS, "+"),
            '-': (TokenType.MINUS, "-"),
            '*': (TokenType.STAR, "*"),
            '/': (TokenType.SLASH, "/"),
            '%': (TokenType.PERCENT, "%"),
            '(': (TokenType.LPAREN, "("),
            ')': (TokenType.RPAREN, ")"),
            '{': (TokenType.LBRACE, "{"),
            '}': (TokenType.RBRACE, "}"),
            ',': (TokenType.COMMA, ","),
            ';': (TokenType.SEMICOLON, ";"),
            ':': (TokenType.COLON, ":"),
        }

        if ch in single_map:
            tok_type, raw = single_map[ch]
            self.tokens.append(Token(tok_type, raw, raw, start_line, start_col, 1))
            return

        raise LexerError(
            f"Unexpected character: {ch!r}",
            line=start_line,
            column=start_col,
            length=1,
            source_line=self._get_current_line(start_line),
        )
